mountpoint-s3/src/prefetch/parquet_prefetch.rs

```diff
@@ -0,0 +1,155 @@
+use crate::sync::Arc;
+use bytes::{Bytes, BytesMut};
+use futures::pin_mut;
+use futures::StreamExt;
+use intervaltree::IntervalTree;
+use mountpoint_s3_client::types::ETag;
+use mountpoint_s3_client::ObjectClient;
+use parquet::errors::ParquetError;
+use parquet::file::footer::decode_footer;
+use parquet::file::metadata::ParquetMetaData;
+use tracing::trace;
+
+use super::PrefetchReadError;
+
+const PARQUET_MAGIC_LEN: usize = 8;
+
+pub async fn read_parquet_metadata<Client>(
+    client: Arc<Client>,
+    bucket: &str,
+    key: &str,
+    if_match: &ETag,
+    total_size: u64,
+) -> Result<(Bytes, std::ops::Range<u64>), PrefetchReadError<Client::ClientError>>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    fetch_metadata(client, bucket, key, if_match, total_size)
+        .await
+        .map(|(metadata, footer, metadata_range)| {
+            let mut combined = BytesMut::with_capacity(metadata.len() + PARQUET_MAGIC_LEN);
+            combined.extend_from_slice(&metadata);
+            combined.extend_from_slice(&footer);
+            (combined.freeze(), metadata_range)
+        })
+        .map_err(|_| PrefetchReadError::MetadataParsingFailed)
+}
+
+async fn fetch_metadata<Client>(
+    client: Arc<Client>,
+    bucket: &str,
+    key: &str,
+    if_match: &ETag,
+    total_size: u64,
+) -> Result<(Bytes, [u8; PARQUET_MAGIC_LEN], std::ops::Range<u64>), ParquetError>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    if total_size < PARQUET_MAGIC_LEN as u64 {
+        return Err(ParquetError::General(
+            "Invalid Parquet file. Size is smaller than footer".to_string(),
+        ));
+    }
+
+    let mut footer = [0_u8; PARQUET_MAGIC_LEN];
+    let footer_range = (total_size - PARQUET_MAGIC_LEN as u64)..total_size;
+
+    fetch_object_part(&client, bucket, key, if_match, footer_range.clone(), |body| {
+        footer.copy_from_slice(&body[..PARQUET_MAGIC_LEN])
+    })
+    .await?;
+
+    let metadata_len = decode_footer(&footer)?;
+    let footer_metadata_len = PARQUET_MAGIC_LEN + metadata_len;
+
+    if footer_metadata_len > total_size as usize {
+        return Err(ParquetError::General(
+            "Invalid Parquet file. Reported metadata length is shorter than expected".to_string(),
+        ));
+    }
```

Comment:
not to be pedantic, but I want to distance from the weird footer naming thing.

```suggestion
    let file_metadata_len = PARQUET_MAGIC_LEN + metadata_len;

    if file_metadata_len > total_size as usize {
        return Err(ParquetError::General(
            "Invalid Parquet file. Reported file metadata length is shorter than expected".to_string(),
        ));
    }
```